File: sdk/ios/Sources/ExperimentationSDK/ExperimentationClient.swift
import Foundation

// MARK: - Protocol

/// Defines the public interface for the ExperimentationSDK client.
///
/// Use this protocol for dependency injection and testing.
public protocol ExperimentationClientProtocol {
    /// Evaluates a feature flag for the given user (decided by the server).
    func evaluateFlag(_ flagKey: String, user: User) async throws -> EvalResult
    /// Retrieves the user's (sticky) experiment assignment from the server.
    func getAssignment(_ experimentKey: String, user: User) async throws -> Assignment
    /// Records a tracking event. Implementations must never throw; the `throws` is kept for
    /// source compatibility only.
    func track(_ event: TrackEvent) async throws
    /// Records several tracking events in one batch request. Never throws; returns `false`
    /// when at least one request failed.
    @discardableResult
    func trackBatch(_ events: [TrackEvent]) async -> Bool
    /// Deprecated no-op kept for source compatibility. Flags are evaluated per user by the
    /// server, so there is nothing to refresh.
    @available(*, deprecated, message: "Flags are evaluated by the server per user; refreshFlags() is a no-op.")
    func refreshFlags() async throws
    /// Clears caches and releases resources.
    func close()
}

public extension ExperimentationClientProtocol {
    /// Default: send every event through ``track(_:)``.
    @discardableResult
    func trackBatch(_ events: [TrackEvent]) async -> Bool {
        for event in events {
            try? await track(event)
        }
        return true
    }

    /// Default: no-op.
    func refreshFlags() async throws {}
}

// MARK: - ExperimentationClient

/// The main entry point for the ExperimentationSDK.
///
/// ## How it works
/// - **The server decides.** `evaluateFlag` calls `GET /api/v1/feature-flags/evaluate/{key}`
///   and `getAssignment` calls `POST /api/v1/tracking/assign`; no bucketing happens on device.
/// - **Caching.** Successful results are cached in memory per user + key for `cacheTTL`
///   seconds (NSCache) and, when `enableOfflineFallback` is on, persisted to `UserDefaults`.
///   Failures are never cached.
/// - **Failure.** On a network or server error the cached value is returned when present;
///   otherwise the call throws (`flagNotFound` / `experimentNotFound` for HTTP 404, which never
///   falls back to a stale persisted value).
/// - **Tracking.** `track` never throws. An event carrying `experimentKey`/`featureFlagKey`
///   goes to `POST /api/v1/tracking/track`; an event without keys is fanned out through
///   `POST /api/v1/tracking/batch` to every cached assignment and evaluated flag of that user
///   (nothing cached → nothing sent).
///
/// ## Thread Safety
/// The client is safe to use from multiple threads and Swift concurrency tasks. Caches are
/// lock-protected and concurrent calls for the same user + key share one in-flight request.
public class ExperimentationClient: ExperimentationClientProtocol {

    /// Maximum events per `POST /api/v1/tracking/batch` request.
    public static let batchLimit = 100

    // MARK: - Properties

    private let config: SdkConfig
    internal var http: HTTPClient
    private let flagCache: UserKeyCache<EvalResult>
    private let assignmentCache: UserKeyCache<Assignment>
    private let offlineStore: OfflineStore

    private var inflightFlags: [String: Task<EvalResult, Error>] = [:]
    private var inflightAssignments: [String: Task<Assignment, Error>] = [:]
    private let inflightLock = NSLock()

    // MARK: - Initializers

    /// Initializes the client with a full configuration object.
    ///
    /// - Parameters:
    ///   - config: The `SdkConfig` specifying base URL, API key, cache settings, etc.
    ///   - session: URLSession used for all requests. Inject a custom session for testing.
    ///   - offlineStore: Persistent store for offline fallback. Defaults to `UserDefaults.standard`
    ///     under the `ep_sdk_` prefix.
    public init(config: SdkConfig, session: URLSession = .shared, offlineStore: OfflineStore? = nil) {
        self.config = config
        self.http = HTTPClient(
            baseURL: config.baseURL,
            apiKey: config.apiKey,
            timeout: config.timeout,
            session: session
        )
        self.flagCache = UserKeyCache(ttl: config.cacheTTL, countLimit: config.cacheSize)
        self.assignmentCache = UserKeyCache(ttl: config.cacheTTL, countLimit: config.cacheSize)
        self.offlineStore = offlineStore ?? OfflineStore()
    }

    /// Convenience initializer for the most common use case.
    ///
    /// - Parameters:
    ///   - baseURL: The API origin (e.g., `"http://localhost:8000"`).
    ///   - apiKey: The API key for authentication.
    public convenience init(baseURL: String, apiKey: String) {
        self.init(config: SdkConfig(baseURL: baseURL, apiKey: apiKey))
    }

    // MARK: - Feature flags

    /// Evaluates a feature flag for the given user via
    /// `GET /api/v1/feature-flags/evaluate/{flagKey}?user_id={user.id}`.
    ///
    /// - Returns: The server's decision (`enabled`) plus the flag's `config`.
    /// - Throws: `ExperimentationError.flagNotFound` when the flag is unknown or not ACTIVE;
    ///   `.networkError` / `.serverError` / `.decodingError` when the request failed and no
    ///   cached value exists.
    public func evaluateFlag(_ flagKey: String, user: User) async throws -> EvalResult {
        if let cached = flagCache.get(userId: user.id, key: flagKey) {
            return cached
        }

        return try await dedupeFlag(key: "\(user.id)\u{0}\(flagKey)") { [self] in
            do {
                let path = "/api/v1/feature-flags/evaluate/\(HTTPClient.encodeComponent(flagKey))"
                let response = try await http.get(path, query: ["user_id": user.id], as: EvalResult.self)
                let result = EvalResult(
                    key: response.key.isEmpty ? flagKey : response.key,
                    enabled: response.enabled,
                    config: response.config
                )
                flagCache.set(userId: user.id, key: flagKey, value: result)
                if config.enableOfflineFallback {
                    offlineStore.saveEvaluation(result, userId: user.id)
                }
                return result
            } catch ExperimentationError.serverError(let code, _) where code == 404 {
                if config.enableOfflineFallback {
                    offlineStore.removeEvaluation(flagKey: flagKey, userId: user.id)
                }
                throw ExperimentationError.flagNotFound(flagKey)
            } catch {
                if config.enableOfflineFallback,
                   let stored = offlineStore.loadEvaluation(flagKey: flagKey, userId: user.id) {
                    return stored
                }
                throw error
            }
        }
    }

    // MARK: - Experiments

    /// Retrieves the user's experiment assignment via `POST /api/v1/tracking/assign`.
    /// The server makes the assignment sticky and records the exposure; `user.attributes`
    /// are sent as `context` for targeting.
    ///
    /// - Throws: `ExperimentationError.experimentNotFound` when the experiment is unknown or not
    ///   ACTIVE; `.networkError` / `.serverError` / `.decodingError` when the request failed and
    ///   no cached value exists.
    public func getAssignment(_ experimentKey: String, user: User) async throws -> Assignment {
        if let cached = assignmentCache.get(userId: user.id, key: experimentKey) {
            return cached
        }

        return try await dedupeAssignment(key: "\(user.id)\u{0}\(experimentKey)") { [self] in
            do {
                let body = AssignRequest(experimentKey: experimentKey, userId: user.id, context: user.attributes)
                let response = try await http.post("/api/v1/tracking/assign", body: body, as: Assignment.self)
                let assignment = Assignment(
                    experimentKey: response.experimentKey.isEmpty ? experimentKey : response.experimentKey,
                    userId: response.userId.isEmpty ? user.id : response.userId,
                    variantId: response.variantId,
                    variantName: response.variantName,
                    isControl: response.isControl,
                    configuration: response.configuration
                )
                assignmentCache.set(userId: user.id, key: experimentKey, value: assignment)
                if config.enableOfflineFallback {
                    offlineStore.saveAssignment(assignment, userId: user.id)
                }
                return assignment
            } catch ExperimentationError.serverError(let code, _) where code == 404 {
                if config.enableOfflineFallback {
                    offlineStore.removeAssignment(experimentKey: experimentKey, userId: user.id)
                }
                throw ExperimentationError.experimentNotFound(experimentKey)
            } catch {
                if config.enableOfflineFallback,
                   let stored = offlineStore.loadAssignment(experimentKey: experimentKey, userId: user.id) {
                    return stored
                }
                throw error
            }
        }
    }

    // MARK: - Tracking

    /// Records an analytics or conversion event. **Never throws** — the `throws` in the
    /// signature is kept for source compatibility; use ``trackWithStatus(_:)`` to learn whether
    /// delivery succeeded.
    ///
    /// - With `experimentKey` / `featureFlagKey`: one `POST /api/v1/tracking/track`.
    /// - Without keys: one `POST /api/v1/tracking/batch` containing one entry per cached
    ///   assignment and one per cached evaluated flag of `event.userId`. Nothing cached → nothing
    ///   is sent.
    public func track(_ event: TrackEvent) async throws {
        _ = await trackWithStatus(event)
    }

    /// Same as ``track(_:)`` but reports the outcome: `true` when every request succeeded
    /// (or there was nothing to send), `false` on any network / server failure.
    @discardableResult
    public func trackWithStatus(_ event: TrackEvent) async -> Bool {
        if event.hasKey {
            do {
                try await http.post("/api/v1/tracking/track", body: event)
                return true
            } catch {
                return false
            }
        }
        return await sendBatches(fanOut(event))
    }

    /// Records several events with `POST /api/v1/tracking/batch` (chunked by
    /// ``batchLimit``). Keyed events are sent as-is; events without keys are fanned out to the
    /// user's cached assignments and flags. Never throws.
    ///
    /// - Returns: `true` when every batch request succeeded (or nothing had to be sent).
    @discardableResult
    public func trackBatch(_ events: [TrackEvent]) async -> Bool {
        let expanded = events.flatMap { $0.hasKey ? [$0] : fanOut($0) }
        return await sendBatches(expanded)
    }

    /// Deprecated no-op. Flags are evaluated per user by the server; there is no flag list to
    /// download. Kept so existing call sites compile.
    @available(*, deprecated, message: "Flags are evaluated by the server per user; refreshFlags() is a no-op.")
    public func refreshFlags() async throws {}

    // MARK: - Cache access

    /// Cached (successful, unexpired) assignments for the user, in the order they were made.
    public func getAssignments(for userId: String) -> [Assignment] {
        assignmentCache.entries(userId: userId).map { $0.value }
    }

    /// Keys of flags successfully evaluated (and still cached) for the user.
    public func getEvaluatedFlags(for userId: String) -> [String] {
        flagCache.entries(userId: userId).map { $0.key }
    }

    /// Drops every in-memory cached evaluation and assignment (the offline store is kept).
    public func clearCache() {
        flagCache.removeAll()
        assignmentCache.removeAll()
    }

    /// Removes everything the SDK persisted to `UserDefaults`.
    public func clearOfflineCache() {
        offlineStore.clearAll()
    }

    /// Clears all in-memory caches and releases resources.
    ///
    /// Call this when the SDK is no longer needed (e.g., on app termination or sign-out).
    /// `OfflineStore` data is not cleared so it remains available on the next launch.
    public func close() {
        clearCache()
        inflightLock.lock()
        inflightFlags.values.forEach { $0.cancel() }
        inflightAssignments.values.forEach { $0.cancel() }
        inflightFlags.removeAll()
        inflightAssignments.removeAll()
        inflightLock.unlock()
    }

    // MARK: - Private helpers

    private struct AssignRequest: Encodable {
        let experimentKey: String
        let userId: String
        let context: [String: AnyCodable]?

        enum CodingKeys: String, CodingKey {
            case experimentKey = "experiment_key"
            case userId = "user_id"
            case context
        }
    }

    private struct BatchRequest: Encodable {
        let events: [TrackEvent]
    }

    /// One entry per cached assignment (`experiment_key`) plus one per cached flag
    /// (`feature_flag_key`) of the event's user.
    private func fanOut(_ event: TrackEvent) -> [TrackEvent] {
        let assignments = getAssignments(for: event.userId).map { event.attributed(experimentKey: $0.experimentKey) }
        let flags = getEvaluatedFlags(for: event.userId).map { event.attributed(featureFlagKey: $0) }
        return assignments + flags
    }

    private func sendBatches(_ events: [TrackEvent]) async -> Bool {
        guard !events.isEmpty else { return true }
        var ok = true
        var index = 0
        while index < events.count {
            let chunk = Array(events[index..<min(index + Self.batchLimit, events.count)])
            do {
                try await http.post("/api/v1/tracking/batch", body: BatchRequest(events: chunk))
            } catch {
                ok = false
            }
            index += Self.batchLimit
        }
        return ok
    }

    /// Share a single in-flight request between concurrent calls for the same user + flag.
    private func dedupeFlag(key: String, _ run: @escaping () async throws -> EvalResult) async throws -> EvalResult {
        let (task, owner) = registerFlagTask(key: key, run)
        defer { if owner { unregisterFlagTask(key: key) } }
        return try await task.value
    }

    /// Share a single in-flight request between concurrent calls for the same user + experiment.
    private func dedupeAssignment(key: String, _ run: @escaping () async throws -> Assignment) async throws -> Assignment {
        let (task, owner) = registerAssignmentTask(key: key, run)
        defer { if owner { unregisterAssignmentTask(key: key) } }
        return try await task.value
    }

    /// Returns the pending task for `key`, or starts one. `owner` is `true` for the caller that
    /// started it (and must unregister it when done). Synchronous so the lock is never held
    /// across a suspension point.
    private func registerFlagTask(
        key: String,
        _ run: @escaping () async throws -> EvalResult
    ) -> (task: Task<EvalResult, Error>, owner: Bool) {
        inflightLock.lock()
        defer { inflightLock.unlock() }
        if let pending = inflightFlags[key] {
            return (pending, false)
        }
        let task = Task { try await run() }
        inflightFlags[key] = task
        return (task, true)
    }

    private func unregisterFlagTask(key: String) {
        inflightLock.lock()
        inflightFlags[key] = nil
        inflightLock.unlock()
    }

    private func registerAssignmentTask(
        key: String,
        _ run: @escaping () async throws -> Assignment
    ) -> (task: Task<Assignment, Error>, owner: Bool) {
        inflightLock.lock()
        defer { inflightLock.unlock() }
        if let pending = inflightAssignments[key] {
            return (pending, false)
        }
        let task = Task { try await run() }
        inflightAssignments[key] = task
        return (task, true)
    }

    private func unregisterAssignmentTask(key: String) {
        inflightLock.lock()
        inflightAssignments[key] = nil
        inflightLock.unlock()
    }
}

File: sdk/dotnet/src/ExperimentationPlatform/ExperimentationClient.cs
using System.Text.Json.Serialization;
using ExperimentationPlatform.Models;

namespace ExperimentationPlatform;

/// <summary>
/// High-level SDK client for evaluating feature flags, getting experiment assignments,
/// and tracking analytics events.
///
/// <para><b>The server decides.</b> <see cref="EvaluateFlagAsync"/> calls
/// <c>GET /api/v1/feature-flags/evaluate/{key}?user_id=…</c> and <see cref="GetAssignmentAsync"/>
/// calls <c>POST /api/v1/tracking/assign</c>; no bucketing happens in the SDK.</para>
///
/// <para><b>Caching.</b> Successful results are cached per user + key for
/// <see cref="SdkConfig.CacheTtlSeconds"/>; failures are never cached. On a network or HTTP
/// failure the cached value is returned when present, otherwise the flag is reported disabled /
/// the assignment is <c>null</c>. None of the public methods throw.</para>
///
/// <para><b>Tracking.</b> An event carrying <c>ExperimentKey</c>/<c>FeatureFlagKey</c> goes to
/// <c>POST /api/v1/tracking/track</c>; an event without keys is fanned out through
/// <c>POST /api/v1/tracking/batch</c> to every cached assignment and evaluated flag of that user
/// (nothing cached → nothing sent).</para>
///
/// <para>Instances are thread-safe and intended to be shared (one per process is typical).</para>
/// </summary>
public class ExperimentationClient : IDisposable
{
    /// <summary>Maximum events per <c>POST /api/v1/tracking/batch</c> request.</summary>
    public const int BatchLimit = 100;

    private readonly ExperimentationHttpClient _httpClient;
    private readonly UserKeyCache<FlagEvaluationResult> _flagCache;
    private readonly UserKeyCache<Assignment> _assignmentCache;
    private bool _disposed;

    /// <summary>
    /// Creates a new client.
    /// </summary>
    /// <param name="config">Required SDK configuration.</param>
    /// <param name="httpClient">
    /// Optional pre-configured <see cref="System.Net.Http.HttpClient"/> for testability.
    /// If not supplied the SDK creates and manages its own instance.
    /// </param>
    public ExperimentationClient(SdkConfig config, System.Net.Http.HttpClient? httpClient = null)
        : this(config ?? throw new ArgumentNullException(nameof(config)), new ExperimentationHttpClient(config, httpClient))
    {
    }

    // Internal constructor for testing with a pre-built http wrapper.
    internal ExperimentationClient(SdkConfig config, ExperimentationHttpClient httpClient)
    {
        _httpClient = httpClient;
        var ttl = TimeSpan.FromSeconds(config.CacheTtlSeconds);
        _flagCache = new UserKeyCache<FlagEvaluationResult>(config.MaxCacheSize, ttl);
        _assignmentCache = new UserKeyCache<Assignment>(config.MaxCacheSize, ttl);
    }

    // -----------------------------------------------------------------
    // Feature flags
    // -----------------------------------------------------------------

    /// <summary>
    /// Evaluates a feature flag for a specific user via
    /// <c>GET /api/v1/feature-flags/evaluate/{flagKey}?user_id={userId}</c>.
    /// </summary>
    /// <param name="flagKey">The feature flag key.</param>
    /// <param name="userId">The user ID.</param>
    /// <param name="attributes">
    /// Accepted for source compatibility; the evaluate endpoint takes no context, so they are not sent.
    /// </param>
    /// <returns>
    /// The server's decision (<c>Key</c>, <c>Enabled</c>, <c>Config</c>). Never throws: on failure
    /// the cached value is returned when present, otherwise a disabled result
    /// (<see cref="FlagEvaluationResult.Disabled"/>). A 404 means the flag is unknown or not ACTIVE.
    /// </returns>
    public async Task<FlagEvaluationResult> EvaluateFlagAsync(
        string flagKey,
        string userId,
        Dictionary<string, object>? attributes = null)
    {
        var cached = _flagCache.Get(userId, flagKey);
        if (cached != null)
            return cached;

        try
        {
            var path = $"/api/v1/feature-flags/evaluate/{Uri.EscapeDataString(flagKey)}";
            var query = new Dictionary<string, string> { ["user_id"] = userId };
            var result = await _httpClient.GetAsync<FlagEvaluationResult>(path, query).ConfigureAwait(false);
            if (string.IsNullOrEmpty(result.Key))
                result.Key = flagKey;

            _flagCache.Set(userId, flagKey, result);
            return result;
        }
        catch (Exception)
        {
            // Never throw — failures are not cached; report the flag as disabled.
            return FlagEvaluationResult.Disabled(flagKey);
        }
    }

    // -----------------------------------------------------------------
    // Experiments
    // -----------------------------------------------------------------

    /// <summary>
    /// Retrieves the user's (sticky) experiment assignment via <c>POST /api/v1/tracking/assign</c>.
    /// The server records the exposure; <paramref name="attributes"/> are sent as <c>context</c>
    /// for targeting rules.
    /// </summary>
    /// <param name="experimentKey">The experiment key.</param>
    /// <param name="userId">The user ID.</param>
    /// <param name="attributes">Optional user attributes (sent as <c>context</c>).</param>
    /// <returns>
    /// The <see cref="Assignment"/>, or <c>null</c> when the experiment is unknown / not ACTIVE
    /// (404) or the request failed and nothing is cached. Never throws.
    /// </returns>
    public async Task<Assignment?> GetAssignmentAsync(
        string experimentKey,
        string userId,
        Dictionary<string, object>? attributes = null)
    {
        var cached = _assignmentCache.Get(userId, experimentKey);
        if (cached != null)
            return cached;

        try
        {
            var body = new AssignRequest
            {
                ExperimentKey = experimentKey,
                UserId = userId,
                Context = attributes
            };
            var assignment = await _httpClient.PostAsync<Assignment>("/api/v1/tracking/assign", body).ConfigureAwait(false);
            if (string.IsNullOrEmpty(assignment.ExperimentKey))
                assignment.ExperimentKey = experimentKey;
            if (string.IsNullOrEmpty(assignment.UserId))
                assignment.UserId = userId;

            _assignmentCache.Set(userId, experimentKey, assignment);
            return assignment;
        }
        catch (Exception)
        {
            return null;
        }
    }

    // -----------------------------------------------------------------
    // Tracking
    // -----------------------------------------------------------------

    /// <summary>
    /// Sends an analytics tracking event. Never throws.
    /// </summary>
    /// <param name="eventName">The event name (experiment metrics match on this name).</param>
    /// <param name="userId">The user who triggered the event.</param>
    /// <param name="properties">Optional event properties (sent as <c>metadata</c>).</param>
    /// <param name="experimentKey">Attribute the event to one experiment.</param>
    /// <param name="featureFlagKey">Attribute the event to one feature flag.</param>
    /// <param name="value">Optional numeric value (revenue, duration, …).</param>
    /// <param name="eventType">Optional event type; defaults to <paramref name="eventName"/>.</param>
    /// <param name="timestamp">Optional event time; sent as ISO-8601 UTC.</param>
    /// <returns>
    /// <c>true</c> when every request succeeded (or there was nothing to send); <c>false</c> on any
    /// network / HTTP error.
    /// </returns>
    /// <remarks>
    /// With <paramref name="experimentKey"/> and/or <paramref name="featureFlagKey"/> one
    /// <c>POST /api/v1/tracking/track</c> is sent. Without keys, one
    /// <c>POST /api/v1/tracking/batch</c> is sent with one entry per cached assignment and one per
    /// cached evaluated flag of the user; if nothing is cached, nothing is sent.
    /// </remarks>
    public Task<bool> TrackAsync(
        string eventName,
        string userId,
        Dictionary<string, object>? properties = null,
        string? experimentKey = null,
        string? featureFlagKey = null,
        double? value = null,
        string? eventType = null,
        DateTimeOffset? timestamp = null)
    {
        return TrackAsync(new TrackEvent
        {
            EventName = eventName,
            UserId = userId,
            Properties = properties,
            ExperimentKey = experimentKey,
            FeatureFlagKey = featureFlagKey,
            Value = value,
            EventType = eventType,
            Timestamp = timestamp
        });
    }

    /// <summary>Sends a prebuilt <see cref="TrackEvent"/> (same rules as the other overload). Never throws.</summary>
    public async Task<bool> TrackAsync(TrackEvent trackEvent)
    {
        if (trackEvent == null)
            return false;

        if (trackEvent.HasKey)
        {
            try
            {
                await _httpClient.PostAsync("/api/v1/tracking/track", trackEvent.ToBody()).ConfigureAwait(false);
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        return await SendBatchesAsync(FanOut(trackEvent)).ConfigureAwait(false);
    }

    /// <summary>
    /// Sends several events with <c>POST /api/v1/tracking/batch</c> (chunked by
    /// <see cref="BatchLimit"/>). Keyed events are sent as-is; events without keys are fanned out to
    /// the user's cached assignments and flags. Never throws.
    /// </summary>
    /// <returns><c>true</c> when every batch request succeeded (or nothing had to be sent).</returns>
    public async Task<bool> TrackBatchAsync(IEnumerable<TrackEvent> events)
    {
        if (events == null)
            return true;

        var expanded = new List<TrackEvent>();
        foreach (var trackEvent in events)
        {
            if (trackEvent == null)
                continue;
            if (trackEvent.HasKey)
                expanded.Add(trackEvent);
            else
                expanded.AddRange(FanOut(trackEvent));
        }

        return await SendBatchesAsync(expanded).ConfigureAwait(false);
    }

    // -----------------------------------------------------------------
    // Cache access
    // -----------------------------------------------------------------

    /// <summary>Cached (successful, unexpired) assignments of the user, in the order they were made.</summary>
    public IReadOnlyList<Assignment> GetAssignments(string userId)
    {
        var result = new List<Assignment>();
        foreach (var entry in _assignmentCache.Entries(userId))
            result.Add(entry.Value);
        return result;
    }

    /// <summary>Keys of flags successfully evaluated (and still cached) for the user.</summary>
    public IReadOnlyList<string> GetEvaluatedFlags(string userId)
    {
        var result = new List<string>();
        foreach (var entry in _flagCache.Entries(userId))
            result.Add(entry.Key);
        return result;
    }

    /// <summary>Drops every cached evaluation and assignment.</summary>
    public void ClearCache()
    {
        _flagCache.Clear();
        _assignmentCache.Clear();
    }

    // -----------------------------------------------------------------
    // Private helpers
    // -----------------------------------------------------------------

    private sealed class AssignRequest
    {
        [JsonPropertyName("experiment_key")]
        public string ExperimentKey { get; set; } = "";

        [JsonPropertyName("user_id")]
        public string UserId { get; set; } = "";

        [JsonPropertyName("context")]
        public Dictionary<string, object>? Context { get; set; }
    }

    private sealed class BatchRequest
    {
        [JsonPropertyName("events")]
        public List<TrackBody> Events { get; set; } = new();
    }

    /// <summary>One entry per cached assignment (experiment_key) plus one per cached flag (feature_flag_key).</summary>
    private List<TrackEvent> FanOut(TrackEvent trackEvent)
    {
        var events = new List<TrackEvent>();
        foreach (var assignment in GetAssignments(trackEvent.UserId))
            events.Add(trackEvent.Attributed(experimentKey: assignment.ExperimentKey));
        foreach (var flagKey in GetEvaluatedFlags(trackEvent.UserId))
            events.Add(trackEvent.Attributed(featureFlagKey: flagKey));
        return events;
    }

    private async Task<bool> SendBatchesAsync(List<TrackEvent> events)
    {
        if (events.Count == 0)
            return true;

        bool ok = true;
        for (int index = 0; index < events.Count; index += BatchLimit)
        {
            var chunk = new BatchRequest();
            int end = Math.Min(index + BatchLimit, events.Count);
            for (int i = index; i < end; i++)
                chunk.Events.Add(events[i].ToBody());

            try
            {
                await _httpClient.PostAsync("/api/v1/tracking/batch", chunk).ConfigureAwait(false);
            }
            catch (Exception)
            {
                ok = false;
            }
        }
        return ok;
    }

    /// <summary>Drops the caches and releases the HTTP client (only when the SDK created it). Safe to call twice.</summary>
    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        ClearCache();
        _httpClient.Dispose();
        GC.SuppressFinalize(this);
    }
}
